File: Problems/0167_Two_Sum_2/Two_Sum_2.py
# coding: utf-8

import os
import sys
import time


class Solution:
    def twoSum0(self, numbers, target):
        """
        :type numbers: List[int]
        :type target: int
        :rtype: List[int]
        """
        resultNumbers = [0]*2
        for i in range(len(numbers)):
            if i > 1 and numbers[i] == numbers[i - 1]:
                continue
            resultNumbers[0] = i + 1
            for j in range(i + 1, len(numbers)):
                if numbers[i] + numbers[j] == target:
                    resultNumbers[1] = j + 1
                    return resultNumbers
                else:
                    if numbers[j] == numbers[j] - 1:
                        break
                    if numbers[i] + numbers[j] > target:
                        break
                    if target > 0 and numbers[j] > target:
                        break

        return resultNumbers

    # two-pointer
    def twoSum1(self, numbers, target):
        l, r = 0, len(numbers)-1
        while l < r:
            s = numbers[l] + numbers[r]
            if s == target:
                return [l+1, r+1]
            elif s < target:
                l += 1
            else:
                r -= 1
    
    # dictionary           
    def twoSum2(self, numbers, target):
        dic = {}
        for i, num in enumerate(numbers):
            if target-num in dic:
                return [dic[target-num]+1, i+1]
            dic[num] = i
    
    # binary search        
    def twoSum3(self, numbers, target):
    #    for i in xrange(len(numbers)):
        for i in range(len(numbers)):
            l, r = i+1, len(numbers)-1
            tmp = target - numbers[i]
            while l <= r:
                mid = l + (r-l)//2
                if numbers[mid] == tmp:
                    return [i+1, mid+1]
                elif numbers[mid] < tmp:
                    l = mid+1
                else:
                    r = mid-1


def array_str_to_int(numbersStr):
    numbers = [0]*len(numbersStr)
    for i in range(len(numbersStr)):
        numbers[i] = int(numbersStr[i])
    return numbers


def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        print("argv[1] = %s" %temp)
        loop_main(temp)


def loop_main(temp):
    flds = temp.split(", ")
    numbersStr = flds[0].replace("[","").replace("]","").split(",")
    numbers = array_str_to_int(numbersStr)
    target = int(flds[1])

    time0 = time.time()

    sl = Solution()
#   result = sl.twoSum(numbers, target)
    result = sl.twoSum3(numbers, target)
    print("result = %s" %result)

    time1 = time.time()

    print("Execute time ... : %f[s]" %(time1 - time0))
    print()


if __name__ == "__main__":
    main()

File: Problems/0167_Two_Sum_2/Two_Sum_2.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int[] TwoSum1(int[] numbers, int target)
    {
        int l = 0, r = numbers.Length - 1;
        int s;
        while (l < r) {
            s = numbers[l] + numbers[r];
            if (s == target)
                return (new int[] {l + 1, r + 1});
            else if (s < target)
                ++l;
            else
                --r;
        }

        return (new int[] {0,0});
    }

    public int[] TwoSum3(int[] numbers, int target)
    {
        for (int i = 0; i < numbers.Length; ++i) {
            int l = i + 1;
            int r = numbers.Length - 1;
            int tmp = target - numbers[i];
            int mid;
            while (l <= r) {
                mid = l + (r - l)/2;
                if (numbers[mid] == tmp)
                    return (new int[] {i+1, mid + 1});
                else if (numbers[mid] < tmp)
                    l = mid + 1;
                else
                    r = mid - 1;
            }
        }

        return (new int[] {0,0});
    }

    public int[] TwoSum_work(int[] numbers, int target)
    {
        int[] resultNumbers = new int [2];
        int i, j;

        for (i = 0; i < numbers.Length; ++i) {
            if (i > 1 && numbers[i] == numbers[i - 1])
                continue;

            resultNumbers[0] = i + 1;

            for (j = i + 1; j < numbers.Length; ++j) {
                if (numbers[i] + numbers[j] == target) {
                    resultNumbers[1] = j + 1;
                    return resultNumbers;
                }
                else {
                    if (numbers[j] == numbers[j - 1])
                        break;
                    if (numbers[i] + numbers[j] > target)
                        break;
                    if (target > 0 && numbers[j] > target)
                        break;
                }
            }
        }

        return resultNumbers;
    }

    private int[] array_str_to_int(string[] workStr)
    {
        int[] numbers = new int[workStr.Length];

        for (int i = 0; i < workStr.Length; i++) {
            numbers[i] = int.Parse(workStr[i]);
        }

        return numbers;
    }

    private string output_array_int(int[] numbers)
    {
        if (numbers.Length == 0)
            return "";

        string resultStr = numbers[0].ToString();
        for (int i = 1; i < numbers.Length; i++)
            resultStr += "," + numbers[i].ToString();

        return resultStr;
    }
	public void Main(string args)
	{
		string[] flds = args.Split(new string[] {", "}, StringSplitOptions.None);
        string[] numbersStr = flds[0].Replace("[","").Replace("]","").Split(',');
        int[] numbers = array_str_to_int(numbersStr);
        int target = int.Parse(flds[1]);

        Console.WriteLine("numbers = [" + output_array_int(numbers) + "], target = " + target);

		System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
		sw.Start();
		
    //  int[] resultNumbers = TwoSum_work(numbers, target);
        int[] resultNumbers = TwoSum1(numbers, target);
    //  int[] resultNumbers = TwoSum3(numbers, target);

		Console.WriteLine("Result = " + output_array_int(resultNumbers));
		
		sw.Stop();
		Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
	}
}
